File: downloadIcloudMail.js
const imaps = require('imap-simple');
const readline = require('readline');
const fs = require('fs'); // To interact with the file system

const IMAP_SERVER = 'imap.mail.me.com';
const IMAP_PORT = 993;

// Reading email and password from command line arguments
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

if (process.argv.length !== 4) {
    console.log("Usage: node downloadIcloudMail.js johndoe@example.com app-specific-password");
    console.log("To make app-specific password:");
    console.log("- go to appleid.apple.com");
    console.log("- Sign-In and Security > App-Specific Passwords.");
    console.log("- create an app-specific password");
    process.exit(1);
}

const EMAIL = process.argv[2];
const PASSWORD = process.argv[3];

const config = {
    imap: {
        user: EMAIL,
        password: PASSWORD,
        host: IMAP_SERVER,
        port: IMAP_PORT,
        tls: true,
        tlsOptions: { rejectUnauthorized: false }
    }
};

async function fetchLatestSentEmail() {
    try {
        const connection = await imaps.connect(config);
        
        await connection.openBox('Sent Messages');

        const searchCriteria = ['ALL'];
        const fetchOptions = { bodies: ['HEADER.FIELDS (FROM SUBJECT DATE)', 'TEXT'] };
        
        const messages = await connection.search(searchCriteria, fetchOptions);

        if (messages.length === 0) {
            console.log("No emails found in the Sent folder.");
            rl.close(); // Close readline and exit
            process.exit(0); // Explicitly end the program
        }

        const latestEmail = messages[messages.length - 1];
        const subject = latestEmail.parts.filter(part => part.which === 'HEADER.FIELDS (FROM SUBJECT DATE)')[0].body.subject;
        const textContent = latestEmail.parts.filter(part => part.which === 'TEXT')[0].body;

        // Ensure subject is treated as a string
        const subjectText = subject ? subject.toString() : "No subject";
        
        console.log(`Subject: ${subjectText}`);
        console.log(`Text Content:\n${textContent}`);

        // Count words in the subject and text content
        const wordCount = (textContent.match(/\w+/g) || []).length + 1 + (subjectText.match(/\w+/g) || []).length;
        console.log(`\n\n\n\nWORD COUNT: ${wordCount}`);

        // Create an object with the email data
        const emailData = {
            subject: subjectText,
            textContent: textContent,
            wordCount: wordCount
        };

        // Save the email data to a JSON file
        fs.writeFileSync('email_data.json', JSON.stringify(emailData, null, 2));

        console.log('Email data saved to email_data.json');

        rl.close(); // Close readline and exit
        process.exit(0); // Explicitly end the program
    } catch (err) {
        console.error("An error occurred:", err);
        rl.close(); // Ensure readline is closed even on error
        process.exit(1); // Exit with error code
    }
}

fetchLatestSentEmail();

File: data_gatherers/iCloud/downloadIcloudMail.py
# from imapclient import IMAPClient
# from pyzmail import PyzMessage
# import sys
# import ssl
# IMAP_SERVER = 'imap.mail.me.com'
# IMAP_PORT = 993

# if(len(sys.argv) != 3):
#     print("usage python3 downloadIcloudMail.py johndoe@example.com app-specific-password\n")
#     print("To make app specific password:\n - go to appleid.apple.com \n - Sign-In and Security > App-Specific Passwords. \n - create an app specific password")
#     sys.exit(1)

# EMAIL = sys.argv[1]
# PASSWORD = sys.argv[2]
# # go to appleid.apple.com
# # "Sign-In and Security" > "App-Specific Passwords."
# # create an app specific password

# def fetch_latest_sent_email():
#     try:
#         context = ssl.create_default_context()
        
#         with IMAPClient(IMAP_SERVER, port=IMAP_PORT, ssl_context=context) as server:
#             server.login(EMAIL, PASSWORD)
#             server.select_folder('Sent Messages')  
            
#             sent_ids = server.search(['ALL'])
#             if not sent_ids:
#                 print("No emails found in the Sent folder.")
#                 return
            
#             latest_email_id = sent_ids[-1]
#             raw_message = server.fetch([latest_email_id], ['BODY[]', 'FLAGS'])
#             message = PyzMessage.factory(raw_message[latest_email_id][b'BODY[]'])
            
#             subject = message.get_subject()
#             text_content = message.text_part.get_payload().decode(message.text_part.charset) if message.text_part else None
#             html_content = message.html_part.get_payload().decode(message.html_part.charset) if message.html_part else None
            
#             print(f"Subject: {subject}")
#             print(f"Text Content:\n{text_content}" if text_content else "No plain text content found.")
#             print(f"HTML Content:\n{html_content}" if html_content else "No HTML content found.")
#             word_count = len(text_content.split(" ")) + 1 + len(subject.split(" "))
#             print(f"\n\n\n\nWORD COUNT: {word_count}")
            
#     except Exception as e:
#         print(f"An error occurred: {e}")

# def list_mailboxes():
#     try:
#         context = ssl.create_default_context()
        
#         with IMAPClient(IMAP_SERVER, port=IMAP_PORT, ssl_context=context) as server:
#             server.login(EMAIL, PASSWORD)
            
#             mailboxes = server.list_folders()
#             print("Available mailboxes:")
#             for mailbox in mailboxes:
#                 print(mailbox)
#     except Exception as e:
#         print(f"An error occurred: {e}")

# fetch_latest_sent_email()